### PR TITLE
Fix more warnings for deprecation and unsafeBitCast()

### PR DESCRIPTION
**Pull request by spevans**
Convert unsafeBitCast() to bindMemory()/withMemoryRebound()

Convert deprecated initialize(from:) to UnsafeMutableBufferPointer.initialize(from:)

### CODE REVIEW

File: Foundation/DateInterval.swift [hunk 1]
```diff
@@ -156,7 +156,10 @@ public struct DateInterval : ReferenceConvertible, Comparable, Hashable {
     public var hashValue: Int {
         var buf: (UInt, UInt) = (UInt(start.timeIntervalSinceReferenceDate), UInt(end.timeIntervalSinceReferenceDate))
         return withUnsafeMutablePointer(to: &buf) {
-            return Int(bitPattern: CFHashBytes(unsafeBitCast($0, to: UnsafeMutablePointer<UInt8>.self), CFIndex(MemoryLayout<UInt>.size * 2)))
+            let count = MemoryLayout<UInt>.size * 2
+            return $0.withMemoryRebound(to: UInt8.self, capacity: count) {
```
Comment (CodaFi) [inline Foundation/DateInterval.swift]: Nit: Repeated use of implicit arguments makes me nervous - if nothing else about readability.  Can you give these names?

Comment (spevans) [inline Foundation/DateInterval.swift]: @CodaFi I've added some names

File: Foundation/NSArray.swift [hunk 2]
```diff
@@ -130,7 +130,7 @@ open class NSArray : NSObject, NSCopying, NSMutableCopying, NSSecureCoding, NSCo
 //        }
         let cnt = array.count
         let buffer = UnsafeMutablePointer<AnyObject>.allocate(capacity: cnt)
-        buffer.initialize(from: optionalArray)
+        _ = UnsafeMutableBufferPointer(start: buffer, count: cnt).initialize(from: optionalArray)
```
Comment (parkera) [inline Foundation/NSArray.swift]: This seems like programming via side-effect. Is there some more explicit way we could express this?

**Comment by parkera** [comment]
That looks much better, thanks. cc @phausler

**Comment by spevans** [comment]
Agreed about the `_ = UnsafeMutableBufferPointer...`
I found a better alternative to `UnsafeMutablePointer.initalize(from:)` which is `UnsafeMutablePointer.initalize(from:count:)`. Not sure why the deprecation warning didnt list it as its not marked at deprecated.

**Comment by spevans** [comment]
@pushkarnk Could you run a test on this PR please

**Comment by alblue** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@phausler Could you have a look at this PR? It fixes up quite a few warnings

File: Foundation/DateInterval.swift [hunk 3]
```diff
@@ -155,8 +155,11 @@ public struct DateInterval : ReferenceConvertible, Comparable, Hashable {
     
     public var hashValue: Int {
         var buf: (UInt, UInt) = (UInt(start.timeIntervalSinceReferenceDate), UInt(end.timeIntervalSinceReferenceDate))
-        return withUnsafeMutablePointer(to: &buf) {
-            return Int(bitPattern: CFHashBytes(unsafeBitCast($0, to: UnsafeMutablePointer<UInt8>.self), CFIndex(MemoryLayout<UInt>.size * 2)))
+        return withUnsafeMutablePointer(to: &buf) { bufferPtr in
+            let count = MemoryLayout<UInt>.size * 2
+            return bufferPtr.withMemoryRebound(to: UInt8.self, capacity: count) { bufferBytes in
```
Comment (phausler) [inline Foundation/DateInterval.swift]: does this potentially allocate?

File: Foundation/DateInterval.swift [hunk 4]
```diff
@@ -155,8 +155,11 @@ public struct DateInterval : ReferenceConvertible, Comparable, Hashable {
     
     public var hashValue: Int {
         var buf: (UInt, UInt) = (UInt(start.timeIntervalSinceReferenceDate), UInt(end.timeIntervalSinceReferenceDate))
-        return withUnsafeMutablePointer(to: &buf) {
-            return Int(bitPattern: CFHashBytes(unsafeBitCast($0, to: UnsafeMutablePointer<UInt8>.self), CFIndex(MemoryLayout<UInt>.size * 2)))
+        return withUnsafeMutablePointer(to: &buf) { bufferPtr in
```
Comment (phausler) [inline Foundation/DateInterval.swift]: we should consider this for the overlay as well

File: Foundation/DateInterval.swift (hunk 3 above)
Comment (spevans) [inline Foundation/DateInterval.swift]: Nothing here seems to indicate an allocation, unless there is something Im not seeing

**Comment by parkera** [comment]
Thanks!

**Pull request merged**